lab-jonathan/app/entry.js

```diff
@@ -4,30 +4,49 @@
 require('./scss/main.scss');
 
 // npm modules
-const cowsay = require('cowsay-browser');
-const angular = require('angular');
-
-// app modules
+let cowsay = require('cowsay-browser');
+let angular = require('angular');
 
 // angular module
-const demoApp = angular.module('demoApp', []);
+let demoApp = angular.module('demoApp', []);
 
 // angular constructus
 demoApp.controller('CowsayController', CowsayController);
 
-function CowsayController($log, $scope){
+function CowsayController($log){
   $log.debug('init CowsayController');
-  let cowsayCtrl = $scope.cowsayCtrl = {};
-  cowsayCtrl.title = 'Moooooo';
-
-  cowsayCtrl.updateCow = function(input){
-    $log.debug('cowsayCtrl.updateCow()');
-    return '\n' + cowsay.say({text: input || 'gimme something to say'});
+  this.title = 'Cowsay Controller Lab';
+  this.history = [];
+
+  cowsay.list((err, cowfiles) => {
+    this.cowfiles = cowfiles;
+    this.currentCow = this.cowfiles[0];
+    console.log('currentCow', this.currentCow);
+  });
+
+  this.updateCow = function(input){
+    $log.debug('this.updateCow()');
+    return '\n' + cowsay.say({text: input || 'gimme something to say', f: this.currentCow});
   };
 
-  cowsayCtrl.helloClick = function(input){
-    $log.debug('cowsayCtrl.helloClick()');
-    $log.log(input);
+  this.speak = function(input){
+    $log.debug('this.updateCow()');
+    this.spoken = this.updateCow(input);
+    this.history.push(this.spoken);
+    console.log('this.history', this.history);
   };
 
+  this.undo = function(){
+    $log.debug('this.undo()');
+    console.log('this.history', this.history);
+    this.history.pop();
```

Comment:
This is broken logic here, given that you're popping off your history array twice. It 'works' but isn't correct. Give it some thought, and play around with it in the browser to see if you can refactor to the correct logic.